src/psyclone/core/access_info.py

```diff
@@ -0,0 +1,364 @@
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author J. Henrichs, Bureau of Meteorology
+# -----------------------------------------------------------------------------
+
+'''This module provides management of variable access information.'''
+
+from __future__ import print_function, absolute_import
+
+from psyclone.core.access_type import AccessType
+
+
+class AccessInfo(object):
+    '''This class stores information about a single access
+    pattern of one variable (e.g. variable is read at a certain location).
+    A location is a number which can be used to compare different accesses
+    (i.e. if one access happens before another). Each consecutive
+    location will have an increasing location number, but read and write
+    accesses in the same statement will have the same location number.
+    If the variable accessed is an array, this class will also store
+    the indices used in the access.
+    Note that the name of the variable is not stored in this instance, it
+    is typically available using VariableAccessInfo class (which stores all
+    AccessInfo objects for a variable).
+
+    :param access: The access type.
+    :type access_type: :py:class:`psyclone.core.access_type.AccessType`
+    :param int location: A number used in ordering the accesses.
+    :param indices: Indices used in the access, defaults to None
+    :type indices: list of :py:class:`psyclone.psyGen.Node` instances \
+        (e.g. Reference, ...)
+    :param node: Node in PSyIR in which the access happens, defaults to None.
+    :type node: :py:class:`psyclone.psyGen.Node` instance.
+
+    '''
+    def __init__(self, access_type, location, node, indices=None):
+        self._location = location
+        self._access_type = access_type
+        self._node = node
+        # Create a copy of the list of indices
+        if indices:
+            self._indices = indices[:]
+        else:
+            self._indices = None
+
+    def change_read_to_write(self):
+        '''This changes the access mode from READ to WRITE.
+        This is used for processing assignment statements,
+        where the LHS is first considered to be READ,
+        and which is then changed to be WRITE.
+
+        :raises InternalError: if the variable originally does not have\
+            READ access.
+        '''
+        if self._access_type != AccessType.READ:
+            from psyclone.psyGen import InternalError
+            raise InternalError("Trying to change variable to 'WRITE' "
```

Comment:
Please could you include the variable name in the error message.

Reply:
*grin* I left it out on purpose, since the AccessInfo object does not have the variable name (it seems a bit of a waste to store the variable name for each access, while de-facto a user will always go via the VariableAccessInfo class, which has the variable name.

So two options:
1) I catch this error in the `VariableAccessInfo` class (since all calls to the `AccessInfo.change_read_to_write` will come from `VariableAccessInfo.change_read_to_write` anyway), and then raise the InternalError with the VariableName.
2) I add the variable name to `AccessInfo`.

My preference would actually be 1) tbh, it seems a waste to add the variable name everywhere ... just more stuff to keep in sync.

Reply:
:-) I noticed it was there in other error messages. I'm happy with option 1 as well.

Reply:
Done.